002-climbingGym

shCHO9801 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->

**AS-IS**
- 공간 데이터 처리 미지원
- ClimbingGym, ClimbingSession 및 UserMonthlyStats 엔티티 미구현
- ClimbingGym, ClimbingSession, UserMonthlyStats 관련 API 미구현
- 페이징 기능 미지원으로 모든 클라이밍 세션을 한 번에 조회하여 성능 저하 가능성
- Jackson과 JTS 'Point' 타입의 직렬화/역직렬화 문제 발생
- Testcontainers 도입 전 H2 데이터베이스 사용으로 인한 테스트 신뢰성 저하
- 통합 테스트 및 유닛 테스트 미작성

**TO-BE**
- Jackson 커스텀 Serializer 및 Deserializer 구현하여 `Point` 타입 직렬화/역직렬화 지원
- ClimbingGym, ClimbingSession, UserMonthlyStats
    - 엔티티 및 Repository 구현
    - Controller 구현하여 CRUD API 제공
    - Service 구현하여 비즈니스 로직 처리
- ClimbingSessionController의 getAllClimbingSessions 메서드에 페이징 기능 추가
- 페이징 기능을 지원하는 ClimbingSessionRepository 메서드 추가
- Testcontainers 도입으로 MySQL 기반의 통합 테스트 환경 구성
- H2와 Testcontainers용 별도 테스트 프로파일(`testContainer`, `testH2`) 설정
- ClimbingGymController 및 ClimbingGymService에 대한 통합 테스트와 유닛 테스트 추가
- Converter 대신 Type 어노테이션 사용하여 모든 엔티티를 CRUD 할 수 있도록 개선
- YearMonth를 LocalDate로 받아 String으로 변환하여 사용하는 방식으로 변경
- UserMonthlyStats에 `getUserMonthlyStatsByYear` 및 `getUserMonthlyStatsPaginated` 메서드 구현 및 페이징 기법 도입


### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
  - Service 유닛테스트
    - ClimbingGymService, ClimbingSessionService, UserMonthlyStats 유닛 테스트 추가
    - ClimbingGymController, ClimbingSessionController, UserMontlyStatsController 통합 테스트 추가
- [x] API 테스트
  - 클라이밍장 생성 및 조회 API 테스트 (페이징 포함)
  - 사용자 월간 통계 조회 API 테스트 (페이징 포함)
  - 중복된 이름으로 클라이밍장 생성 시 에러 처리 테스트
  - 페이징 파라미터를 사용한

shCHO9801 committed:
feat : ClimbingGym 관련 코드 구현
- **ClimbingGym 엔티티 생성**
  - `id`, `name`, `location` (Point 타입), `price`, `parkingInfo`, `difficultyChart`, `amenities`, `createdAt`, `updatedAt` 필드 추가
  - `location` 필드를 MySQL의 `POINT` 타입으로 매핑

- **서비스 레이어 구현**
  - `ClimbingGymService` 클래스 생성
  - 클라이밍장 생성, 조회, 목록 조회 기능 구현
  - 중복 이름 검증 로직 추가

- **컨트롤러 레이어 구현**
  - `ClimbingGymController` 클래스 생성
  - RESTful API 엔드포인트 (`/api/gyms`) 추가
  - 클라이밍장 생성 및 조회 요청 처리

- **커스텀 Serializer 및 Deserializer 구현**
  - `PointSerializer` 클래스 생성하여 `Point` 객체를 GeoJSON 형식으로 직렬화
  - `PointDeserializer` 클래스 생성하여 GeoJSON 형식의 JSON 데이터를 `Point` 객체로 역직렬화

- **Jackson 설정 추가**
  - `JacksonConfig.java` 수정
  - 커스텀 Serializer 및 Deserializer를 Jackson `ObjectMapper`에 등록

shCHO9801 committed:
test : ClimbingGymController 및 Service 인테그레이션 및 유닛 테스트
- 통합 테스트
  - Testcontainers와 MySQL을 사용한 ClimbingGymControllerTest 작성
  - Point 타입 데이터의 직렬화 및 역직렬화 검증
  - 클라이밍장 생성 및 조회 기능 테스트

- 유닛 테스트
  - Mockito를 사용한 ClimbingGymServiceTest 작성
  - 서비스 로직 검증 및 중복 이름 처리 테스트

- 테스트 프로필 분리
  - H2와 Testcontainers용 별도 프로파일 설정 추가

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/climbinggym/controller/ClimbingGymController.java:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  @GetMapping
+  public ResponseEntity<List<GetGym>> getAllClimbingGyms() {
```
paging 처리를 하면 좋겠습니다. 클라이밍장의 개수가 많으면 한번에 가지고 오기에는 부하가 많이 걸릴 것으로 보입니다

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/climbinggym/entity/ClimbingGym.java:
```diff
@@ -0,0 +1,66 @@
+  private String parkingInfo;
+
+  @Column(name = "difficulty_chart", columnDefinition = "JSON")
+  @Convert(converter = JsonConverter.class)
```
별도의 converter를 선언하는 것 보단 Type을 선언하여 사용하는 것을 추천드립니다. 
https://unluckyjung.github.io/db/2023/06/04/mysql-json-type/

mingming-mentor replied:
현재 구현하신대로라면 MySQL의 json type을 쓰는것이 아니라 어플리케이션에서 string으로 변환하여 넣어주는 형태입니다. 
MySQL에서 제공하는 json type column의 json list를 쓰는 것이 추후 search 등에 있어서 index를 걸수 있는 등 훨씬 효율적으로 동작할 수 있기 때문에 이런 부분들을 공부해보시고 적용하시면 좋겠어요

shCHO9801 replied:
@mingming-mentor 멘토님 피드백 감사합니다 다른 부분들은 진행해 볼 수 있을 것 같은데 말씀해 주신 Type 부분에 관해서 질문드리고싶은 부분이 있습니다
hibernate 6 이상 부터는 @TypeDef를 사용할 수 없어서 AttributuConvert를 사용을 권고한다는 내용을 확인해서 hibernate의 버전을 낮추고 해당 부분을 낮추고 말씀해주신 부분 구현을 진행하면 괜찮을까요??

mingming-mentor replied:
@shCHO9801 님, hibernate 6 이상이라 @TypeDef 사용이 어렵다면 @JdbcTypeCode 를 사용하셔도 좋을 것 같습니다.

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/climbinggym/entity/ClimbingGym.java:
```diff
@@ -0,0 +1,66 @@
+  @PrePersist
+  protected void onCreate() {
+    this.createdAt = LocalDateTime.now();
+    this.updatedAt = LocalDateTime.now();
```
생성시에 updatedAt도 같이 넣어주는 이유가 있나요? createdAt만 설정해주는 것이 좀더 낫지 않나 하여 여쭤보아요

shCHO9801 replied:
넵 빈 칼럼으로 두고싶지 않아서 생성시에 초기화 줄 값 대신 사용한다는 생각으로 작성해뒀었는데 삭제 처리 하고 updatedAt을 null이 가능하도록 세팅 하겠습니다

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/climbinggym/service/ClimbingGymService.java:
```diff
@@ -0,0 +1,73 @@
+
+  private final ClimbingGymRepository climbingGymRepository;
+
+  private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
```
static도 같이 선언하시면 좋겠습니다

mingming-mentor reviewed (approved):
수고 많으셨습니다

shCHO9801 committed:
feat : 유저 운동 세션 구현
엔티티(Entity):
ClimbingSession 엔티티: 필요한 필드(id, date, duration, difficultyLevelsCompleted, user, climbingGym 등)와 JPA 어노테이션 추가.
UserMonthlyStats 엔티티: 사용자 월간 통계를 위한 필드(id, userNum, yearMonth, totalDuration 등)와 JPA 어노테이션 추가.
DTO(Data Transfer Object):
CreateSession.Request 및 CreateSession.Response 클래스: 클라이밍 세션 생성 및 응답을 위한 필드 정의.
UserMonthlyStats 관련 DTO 클래스: 필요 시 추가.
레포지토리(Repository):
ClimbingSessionRepository: ClimbingSession 엔티티를 위한 JPA 레포지토리 인터페이스 생성.
UserMonthlyStatsRepository: UserMonthlyStats 엔티티를 위한 JPA 레포지토리 인터페이스 생성.
서비스(Service):
ClimbingSessionService: 클라이밍 세션 관련 비즈니스 로직 구현.
UserMonthlyStatsService: 사용자 월간 통계 관련 비즈니스 로직 구현.
컨트롤러(Controller):
ClimbingSessionController: 클라이밍 세션 관련 API 엔드포인트 (POST /api/climbing-session, GET /api/climbing-session) 구현.
UserMonthlyStatsController: 사용자 월간 통계 관련 API 엔드포인트 구현 (필요 시 추가).
기타:
필요한 AttributeConverter 클래스(YearMonthConverter, JsonMapConverter 등) 구현 및 설정.
JacksonConfig 클래스 수정하여 JavaTimeModule 등록 및 SerializationFeature.WRITE_DATES_AS_TIMESTAMPS 비활성화.

shCHO9801 committed:
test : 인테그레이션 테스트, 로직 테스트 작성
테스트 클래스:
ClimbingSessionServiceTest: ClimbingSessionService의 메서드들을 대상으로 한 단위 테스트 작성.
UserMonthlyStatsServiceTest: UserMonthlyStatsService의 메서드들을 대상으로 한 단위 테스트 작성.

shCHO9801 committed:
feat : 피드백 반영
- getAllGyms 메소드 페이징 기법 사용
- updatedAt 생성시 초기화 삭제

shCHO9801 committed:
fix : 테스트별 코드 수정에 의한 오류 수정

shCHO9801 committed:
feat: Type 어노테이션 사용하여 모든 엔티티 CRUD 기능 구현

- Converter 대신 Type 어노테이션을 사용하여 ClimbingGym, ClimbingSession, UserMonthlyStats 등의 엔티티에 대한 CRUD 기능을 구현
YearMonth 대신 LocalDate 사용 및 String 변환 로직 추가

- YearMonth를 사용하던 부분을 LocalDate로 변경하고, 이를 String으로 변환하여 사용하는 로직을 추가

shCHO9801 committed:
feat : ClimbingSessionController에 페이징 기능 도입하여 getAllClimbingSessions 메서드 개선
- getAllClimbingSessions 메서드를 List 대신 Paging을 이용하여 호출할 수 있도록 개선

shCHO9801 committed:
feat : UserMonthlyStatsService에 getUserMonthlyStatsByYear 및 getUserMonthlyStatsPaginated 메서드 구현 및 페이징 적용
- UserMonthlyStats 서비스에 getUserMonthlyStatsByYear와 getUserMonthlyStatsPaginated 메서드를 구현하고, 페이징 기법을 적용

shCHO9801 committed:
feat : 인티그레이션 및 유닛 테스트 추가

shCHO9801 committed:
Merge branch 'main' into 002-climbing
